File: solana_toolbox_web3/src/ToolboxIdlPath.ts
type ToolboxIdlPathPartDiscriminant = 'empty' | 'index' | 'key';
type ToolboxIdlPathPartContent = null | bigint | string;

export class ToolboxIdlPathPart {
  private discriminant: ToolboxIdlPathPartDiscriminant;
  private content: ToolboxIdlPathPartContent;

  private constructor(
    discriminant: ToolboxIdlPathPartDiscriminant,
    content: ToolboxIdlPathPartContent,
  ) {
    this.discriminant = discriminant;
    this.content = content;
  }
  public static empty(): ToolboxIdlPathPart {
    return new ToolboxIdlPathPart('empty', null);
  }
  public static index(value: bigint): ToolboxIdlPathPart {
    return new ToolboxIdlPathPart('index', value);
  }
  public static key(value: string): ToolboxIdlPathPart {
    return new ToolboxIdlPathPart('key', value);
  }

  public isEmpty(): boolean {
    return this.discriminant === 'empty';
  }

  public key(): string | undefined {
    if (this.discriminant === 'key') {
      return this.content as string;
    }
    return undefined;
  }
  public index(): bigint | undefined {
    if (this.discriminant === 'index') {
      return this.content as bigint;
    }
    return undefined;
  }

  public value(): string {
    switch (this.discriminant) {
      case 'empty':
        return '';
      case 'index':
        return (this.content as bigint).toString();
      case 'key':
        return this.content as string;
    }
  }

  public traverse<P1, P2, T>(
    visitor: {
      empty: (value: null, p1: P1, p2: P2) => T;
      index: (value: bigint, p1: P1, p2: P2) => T;
      key: (value: string, p1: P1, p2: P2) => T;
    },
    p1: P1,
    p2: P2,
  ): T {
    switch (this.discriminant) {
      case 'empty':
        return visitor.empty(this.content as null, p1, p2);
      case 'index':
        return visitor.index(this.content as bigint, p1, p2);
      case 'key':
        return visitor.key(this.content as string, p1, p2);
    }
  }
}

export class ToolboxIdlPath {
  private readonly parts: Array<ToolboxIdlPathPart>;

  private constructor(parts: Array<ToolboxIdlPathPart>) {
    this.parts = parts;
  }

  public static tryParse(path: string): ToolboxIdlPath {
    const parts: Array<ToolboxIdlPathPart> = [];
    for (const part of path.split('.')) {
      if (part === '') {
        parts.push(ToolboxIdlPathPart.empty());
      } else if (/^\d+$/.test(part)) {
        parts.push(ToolboxIdlPathPart.index(BigInt(part)));
      } else {
        parts.push(ToolboxIdlPathPart.key(part));
      }
    }
    if (parts[0]!.isEmpty()) {
      parts.shift();
    }
    return new ToolboxIdlPath(parts);
  }

  public splitFirst():
    | { first: ToolboxIdlPathPart; rest: ToolboxIdlPath }
    | undefined {
    if (this.parts.length === 0) {
      return undefined;
    }
    return {
      first: this.parts[0]!,
      rest: new ToolboxIdlPath(this.parts.slice(1)),
    };
  }

  public isEmpty(): boolean {
    return this.parts.length === 0;
  }

  public value(): string {
    return this.parts.map((p) => p.value()).join('.');
  }
}

File: solana_toolbox_idl/src/toolbox_idl_path.rs
use anyhow::Result;

#[derive(Debug, Clone, PartialEq)]
pub struct ToolboxIdlPath {
    pub parts: Vec<ToolboxIdlPathPart>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum ToolboxIdlPathPart {
    Empty,
    Index(usize),
    Key(String),
}

impl ToolboxIdlPath {
    pub fn try_parse(value: &str) -> Result<ToolboxIdlPath> {
        let mut parts = vec![];
        for part in value.split(".") {
            if part.is_empty() {
                parts.push(ToolboxIdlPathPart::Empty)
            } else if part.chars().all(|c| c.is_ascii_digit()) {
                parts.push(ToolboxIdlPathPart::Index(part.parse()?))
            } else {
                parts.push(ToolboxIdlPathPart::Key(part.to_string()))
            }
        }
        if parts[0] == ToolboxIdlPathPart::Empty {
            parts.remove(0);
        }
        Ok(ToolboxIdlPath { parts })
    }

    pub fn concat(&self, other: &ToolboxIdlPath) -> ToolboxIdlPath {
        ToolboxIdlPath {
            parts: self
                .parts
                .iter()
                .chain(other.parts.iter())
                .cloned()
                .collect::<Vec<_>>(),
        }
    }

    pub fn is_empty(&self) -> bool {
        self.parts.is_empty()
    }

    pub fn split_first(&self) -> Option<(ToolboxIdlPathPart, ToolboxIdlPath)> {
        if let Some((first, rest)) = self.parts.split_first() {
            return Some((
                first.clone(),
                ToolboxIdlPath {
                    parts: rest.to_vec(),
                },
            ));
        }
        None
    }

    pub fn value(&self) -> String {
        let mut parts = vec![];
        for part in &self.parts {
            parts.push(part.value());
        }
        parts.join(".")
    }
}

impl ToolboxIdlPathPart {
    pub fn key(&self) -> Option<&str> {
        match self {
            ToolboxIdlPathPart::Empty => None,
            ToolboxIdlPathPart::Key(key) => Some(key),
            ToolboxIdlPathPart::Index(_) => None,
        }
    }

    pub fn index(&self) -> Option<usize> {
        match self {
            ToolboxIdlPathPart::Empty => None,
            ToolboxIdlPathPart::Key(_) => None,
            ToolboxIdlPathPart::Index(index) => Some(*index),
        }
    }

    pub fn value(&self) -> String {
        match self {
            ToolboxIdlPathPart::Empty => "".to_string(),
            ToolboxIdlPathPart::Key(key) => key.to_string(),
            ToolboxIdlPathPart::Index(index) => index.to_string(),
        }
    }
}
